docs: introduce metric engine in contributor guide (#843)
GreptimeTeam/docs: docs/v0.7/en/contributor-guide/datanode/metric-engine.md

# Metric Engine

## Overview

> The `Metric` engine is a component of GreptimeDB, and it's an implementation of the "storage engine". It mainly targets scenarios with a large number of small tables for observable metrics.

Reviewer: ```suggestion
The `Metric` engine is a component of GreptimeDB, and it's an implementation of the _storage engine_. It mainly targets scenarios with a large number of small tables for observable metrics.
```

And others.

Or @nicecui has different docs guidelines >_<

Reviewer 2: Just the `storage engine` is ok. we do not need `"storage engine"` or `_storage engine_`.
If the product name can be translated into other languages, then use the name in plain format without emphasis.
This will save the author most of the energy to focus on the content logic of the feature.

> The `Metric` engine is a component of GreptimeDB, and it's an implementation of the "storage engine". It mainly targets scenarios with a large number of small tables for observable metrics.

Reviewer 2: ```suggestion
The `Metric` engine is a component of GreptimeDB, and it's an implementation of the storage engine. It mainly targets scenarios with a large number of small tables for observable metrics.
```

Its main feature is to use synthetic physical wide tables to store a large amount of small table data, achieving effects such as reuse of the same column and metadata. This reduces storage overhead for small tables and improves columnar compression efficiency. The concept of a table becomes even more lightweight under the `Metric` engine.

## Concepts

> The `Metric` engine introduces two new concepts: "logical table" and "physical table". From the user's perspective, logical tables are exactly like ordinary ones. From a storage point-of-view, physical Regions are just regular Regions.

Reviewer: ```suggestion
The `Metric` engine introduces two new concepts: "logical table" and "physical table". From the user's perspective, logical tables are exactly like ordinary ones.
```

We talk about table. Why suddenly "physical Regions". Is there "logical regions"?

### Physical Table

A physical table is a table that actually stores data, possessing several physical Regions defined by partition rules.

## Architecture and Design

The main design architecture of the `Metric` engine is as follows:

![Arch](/metric-engine-arch.png)

In the current version implementation, the `Metric` engine reuses the `Mito` engine to achieve storage and query capabilities for physical data. It also provides access to both physical tables and logical tables simultaneously.

Regarding partitioning, logical tables have identical partition rules and Region distribution as physical tables. This makes sense because the data of logical tables are directly stored in physical tables, so their partition rules are consistent.

Concerning routing metadata, the routing address of a logical table is a logical address - what its corresponding physical table is - then through this physical table for secondary routing to obtain the real physical address. This indirect routing method can significantly reduce the number of metadata modifications required when Region migration scheduling occurs in Metric engines.

> Operationally speaking, The `Metric` engine only supports limited operations on physical tables to prevent misoperations such as prohibiting writing into or deleting from a  physical table which could affect user's logic-table data. Generally speaking, users can consider that they have read-only access to these Physical Tables.

Reviewer 2: ```suggestion
Operationally speaking, The `Metric` engine only supports limited operations on physical tables to prevent misoperations such as prohibiting writing into or deleting from a physical table which could affect user's logic-table data. Generally speaking, users can consider that they have read-only access to these physical tables.
```

> To improve performance during simultaneous DDL (Data Definition Language) operations on many tables like automatic creation requests brought about by large amounts of metrics during Prometheus Remote Write cold start-up or modification requests for numerous route-tables mentioned earlier during migration of many Physical Regions; The 'Metric' Engine has introduced some batch DDL operations. These batch DDL operations can merge lots of DDL actions into one request thereby reducing queries and modifications times for Metadata thus enhancing performance.

Reviewer 2: ```suggestion
To improve performance during simultaneous DDL (Data Definition Language) operations on many tables, the 'Metric' engine has introduced some batch DDL operations. These batch DDL operations can merge lots of DDL actions into one request thereby reducing queries and modifications times for metadata thus enhancing performance. This feature is particularly beneficial in scenarios such as the automatic creation requests brought about by large amounts of metrics during Prometheus Remote Write cold start-up, as well as the modification requests for numerous route-tables mentioned earlier during migration of many physical regions.
```

> Apart from Physical Data Regions belonging to Physical Tables; The 'Metric' Engine creates an additional Metadata Region physically for each individual Physical Data Region used in storing some Metadata needed by itself while maintaining mapping and other states. This Metadata includes the mapping relationship between Logical Tables and Physical Tables, the mapping relationship between Logical Columns and Physical Columns etc.

Reviewer 2: ```suggestion
Apart from physical data regions belonging to physical tables, the 'Metric' engine creates an additional metadata region physically for each individual physical data region used in storing some metadata needed by itself while maintaining mapping and other states. This metadata includes the mapping relationship between logical tables and physical tables, the mapping relationship between logical columns and physical columns etc.
```
